lib/rules/prefer-arrow-callback.js

```diff
@@ -246,7 +249,37 @@ module.exports = {
                     !scopeInfo.super &&
                     !scopeInfo.meta
                 ) {
-                    context.report(node, "Unexpected function expression.");
+                    context.report({
+                        node,
+                        message: "Unexpected function expression.",
+                        fix(fixer) {
+
+                            // FIXME: Is there a better way to detect parser support for arrow functions?
+                            if (!context.parserOptions || !context.parserOptions.ecmaVersion || context.parserOptions.ecmaVersion < 6) {
+
+                                // If ES6 parsing is not enabled, creating arrow functions will cause syntax errors, so don't do that.
+                                return null;
+                            }
+
+                            if (!callbackInfo.isLexicalThis && scopeInfo.this) {
+
+                                // If the callback function does not have .bind(this) and contains a reference to `this`, there
+                                // is no way to determine what `this` should be, so don't perform any fixes.
+                                return null;
```

Comment:
I guess this is cases that `allowUnboundThis` option is enabled.
I agree that we should not modify the code in this case.